src/headers.ts

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as core from '@actions/core';
+
+const customMetadataPrefix = 'x-goog-meta-';
+
+interface CustomMetadata {
+  [key: string]: string;
+}
+
+export interface Metadata {
+  cacheControl?: string;
+  contentDisposition?: string;
+  contentEncoding?: string;
+  contentLanguage?: string;
+  contentType?: string;
+  customTime?: string;
+  metadata?: CustomMetadata;
+}
+
+function parseSingleHeaderLine(input: string): [string, string] {
+  const parts = input.split(':');
+  const name = parts[0].trim().toLowerCase();
+  const value = parts
+    .splice(1)
+    .join(':')
+    .trim();
+  return [name, value];
+}
+
+function parseHeaderLines(input: string): Map<string, string> {
+  const headerPairs = input
+    .split(/\r?\n/)
+    .filter((line) => line.trim())
+    .map((line) => parseSingleHeaderLine(line));
+  return new Map<string, string>(headerPairs);
+}
+
+/**
+ * Parses multiline headers input to the expected metadata object
+ * by the GCS library.
+ *
+ * Custom metadata must be prefixed with `x-goog-meta-`. Invalid
+ * headers are ignored and logged as warnings.
+ *
+ * @param input multiline string with headers.
+ * @returns The parsed metadata object.
+ */
+export function parseHeadersInput(input: string): Metadata {
+  const headers = parseHeaderLines(input);
+  const metadata: Metadata = {};
+  headers.forEach((value, key) => {
+    if (key.startsWith(customMetadataPrefix)) {
+      if (!metadata.metadata) {
+        metadata.metadata = {};
+      }
+      metadata.metadata[key.slice(customMetadataPrefix.length)] = value;
+    } else {
+      switch (key) {
+        case 'cache-control':
+          metadata.cacheControl = value;
+          break;
+        case 'content-disposition':
+          metadata.contentDisposition = value;
+          break;
+        case 'content-encoding':
+          metadata.contentEncoding = value;
+          break;
+        case 'content-language':
+          metadata.contentLanguage = value;
+          break;
+        case 'content-type':
+          metadata.contentType = value;
+          break;
+        case 'custom-time':
+          metadata.customTime = value;
```

Comment:
I was wondering if there exists a mapping of this in the upstream lib so we don't have to maintain this.

Reply:
Upstream `@google-cloud/storage` expects the structured metadata object like [this](https://github.com/googleapis/nodejs-storage/blob/e476961ea621b2016069f9de7f3786076b37c78a/samples/fileSetMetadata.js#L44-L55) and it does not support "headers". I'm afraid there's no mapping defined there but I'll double check it. I assume there might be something like this in gsutil source though.

To sum up, I think we've got 2 options:

**1) Directly map metadata values (so we won't need to write this mapping).**

However, then we need to have multiple inputs instead of a single `headers` in GH action, e.g.

```yaml
gzip: false
...
contentType: application/json
cacheControl: public, max-age=3600
metadata: |-
  custom-field: custom-value
```

GH action inputs cannot contain nested values so it's not possible to simply write it like this:
```yaml
gzip: false
...
metadata:
  contentType: application/json
  cacheControl: public, max-age=3600
    metadata:
      custom-field: custom-value
```

We could encode it as a JSON/YAML but I'm not sure that's the right way to go.

**2) Provide "headers" interface and map these values to the expected structure by `@google-cloud/storage`.**

That's the approach I used in current implementation. It follows the feature request from #22 and it's similar to the way people know it from gsutil.

Reply:
Thanks I think we can keep the current implementation for now